# Carlos Llompart López
Inteligencia Artificial

Práctica 3. FIFA practice

Miquel Miró Nicolau

https://github.com/cllompart/PRACTICA-FIFA



# Fifa practice

We want to predict the *value* of the players using the information of the player.

## Import libraries

First of all we import the libraries we're gonna use.

## Read the data

To read the data we're gonna use ``pandas`` and an utility from the library
``os``.


In [1]:
import os

from sklearn.model_selection import train_test_split
from sklearn import linear_model
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

Creamos una base que contrendrá la constrrucción del dataset con los datos de entrenamiento relativos a la variable objetivo 'Valor'. De esta manera, comprobaremos como mejoran esos datos en nuestro modelo de predicción.

In [2]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))
base = df
base.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

Ya tenemos los jugadores cargados en el data frame

Now we select one columns. The result will be an object from the ``pandas`` library a ``series`` object. This object can be converted to a more standard ``numpy`` array.

In [3]:
name = base["Name"]
name 

0                  L. Messi
1         Cristiano Ronaldo
2                 Neymar Jr
3                    De Gea
4              K. De Bruyne
                ...        
18202          J. Lundstram
18203    N. Christoffersson
18204             B. Worman
18205        D. Walker-Rice
18206             G. Nugent
Name: Name, Length: 18207, dtype: object

## Data analysis 

First of all we want to known what is happening to the data. To do so we increase the number of columns to show. After that we're gonna use a special instruction that will show us the data structure.



In [4]:
pd.set_option('display.max_columns', None)

In [5]:
base.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

In [6]:
base.describe()

Unnamed: 0             ID           Age       Overall     Potential  \
count  18207.000000   18207.000000  18207.000000  18207.000000  18207.000000   
mean    9103.000000  214298.338606     25.122206     66.238699     71.307299   
std     5256.052511   29965.244204      4.669943      6.908930      6.136496   
min        0.000000      16.000000     16.000000     46.000000     48.000000   
25%     4551.500000  200315.500000     21.000000     62.000000     67.000000   
50%     9103.000000  221759.000000     25.000000     66.000000     71.000000   
75%    13654.500000  236529.500000     28.000000     71.000000     75.000000   
max    18206.000000  246620.000000     45.000000     94.000000     95.000000   

            Special  International Reputation     Weak Foot   Skill Moves  \
count  18207.000000              18159.000000  18159.000000  18159.000000   
mean    1597.809908                  1.113222      2.947299      2.361308   
std      272.586016                  0.394031      0.660456      0.756164   
min      731.000000                  1.000000      1.000000      1.000000   
25%     1457.000000                  1.000000      3.000000      2.000000   
50%     1635.000000                  1.000000      3.000000      2.000000   
75%     1787.000000                  1.000000      3.000000      3.000000   
max     2346.000000                  5.000000      5.000000      5.000000   

       Jersey Number      Crossing     Finishing  HeadingAccuracy  \
count   18147.000000  18159.000000  18159.000000     18159.000000   
mean       19.546096     49.734181     45.550911        52.298144   
std        15.947765     18.364524     19.525820        17.379909   
min         1.000000      5.000000      2.000000         4.000000   
25%         8.000000     38.000000     30.000000        44.000000   
50%        17.000000     54.000000     49.000000        56.000000   
75%        26.000000     64.000000     62.000000        64.000000   
max        99.000000     93.000000     95.000000        94.000000   

       ShortPassing       Volleys     Dribbling         Curve    FKAccuracy  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      58.686712     42.909026     55.371001     47.170824     42.863153   
std       14.699495     17.694408     18.910371     18.395264     17.478763   
min        7.000000      4.000000      4.000000      6.000000      3.000000   
25%       54.000000     30.000000     49.000000     34.000000     31.000000   
50%       62.000000     44.000000     61.000000     48.000000     41.000000   
75%       68.000000     57.000000     68.000000     62.000000     57.000000   
max       93.000000     90.000000     97.000000     94.000000     94.000000   

        LongPassing   BallControl  Acceleration   SprintSpeed       Agility  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      52.711933     58.369459     64.614076     64.726967     63.503607   
std       15.327870     16.686595     14.927780     14.649953     14.766049   
min        9.000000      5.000000     12.000000     12.000000     14.000000   
25%       43.000000     54.000000     57.000000     57.000000     55.000000   
50%       56.000000     63.000000     67.000000     67.000000     66.000000   
75%       64.000000     69.000000     75.000000     75.000000     74.000000   
max       93.000000     96.000000     97.000000     96.000000     96.000000   

          Reactions       Balance     ShotPower       Jumping       Stamina  \
count  18159.000000  18159.000000  18159.000000  18159.000000  18159.000000   
mean      61.836610     63.966573     55.460047     65.089432     63.219946   
std        9.010464     14.136166     17.237958     11.820044     15.894741   
min       21.000000     16.000000      2.000000     15.000000     12.000000   
25%       56.000000     56.000000     45.000000     58.000000     56.000000   
50%       62.000000     66.000000     59.000000     66.000000     66.00000

We can see that the two first columns do not provide any meaningful information so we can delete them.

In [7]:
base = base.iloc[:, 2:]
base

Name  Age  \
0                L. Messi   31   
1       Cristiano Ronaldo   33   
2               Neymar Jr   26   
3                  De Gea   27   
4            K. De Bruyne   27   
...                   ...  ...   
18202        J. Lundstram   19   
18203  N. Christoffersson   19   
18204           B. Worman   16   
18205      D. Walker-Rice   17   
18206           G. Nugent   16   

                                                Photo Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png     Belgium   
...                                               ...         ...   
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
0      https://cdn.sofifa.org/flags/52.png       94         94   
1      https://cdn.sofifa.org/flags/38.png       94         94   
2      https://cdn.sofifa.org/flags/54.png       92         93   
3      https://cdn.sofifa.org/flags/45.png       91         93   
4       https://cdn.sofifa.org/flags/7.png       91         92   
...                                    ...      ...        ...   
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                      Club                                       Club Logo  \
0             FC Barcelona    https://cdn.sofifa.org/teams/2/light/241.png   
1                 Juventus     https://cdn.sofifa.org/teams/2/light/45.png   
2      Paris Saint-Germain     https://cdn.sofifa.org/teams/2/light/73.png   
3        Manchester United     https://cdn.sofifa.org/teams/2/light/11.png   
4          Manchester City     https://cdn.sofifa.org/teams/2/light/10.png   
...                    ...                                             ...   
18202      Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203       Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204     Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206      Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

         Value   Wage  Special Preferred Foot  International Reputation  \
0      €110.5M  €565K     2202           Left                       5.0   
1         €77M  €405K     2228          Right                       5.0   
2      €118.5M  €290K     2143          Right                       5.0   
3         €72M  €260K     1471          Right                       4.0   
4        €102M  €355K     2281          Right                       4.0   
...        ...    ...      ...            ...                       ...   
18202     €60K    €1K     1307          Right                       1.0   
18203     €60K    €1K     1098          Right                       1.0   
18204     €60K    €1K     1189          Right                       1.0   
18205     €60K    €1K     1228          Right                       1.0   
18206     €60K    €1K     1321          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate   Body Type Real Face Position  \
0            4.0          4.0  Medium/ Medium       Messi       Yes     

Con la funcion describe(), fijándonos en la desviación típica, podemos comprobar que existen datos que no nos proporcionan información importante para evaluar el valor de un jugador. Esta información solo provocará que empeore nuestro modelo de entrenamiento Por lo tanto, eliminaremos las columnas : Nombre, Foto, Bandera, Logo del Club,Cara Real y Dorsal.

In [8]:
base.drop(['Name','Photo','Flag','Club Logo','Real Face','Jersey Number'], axis = 1,inplace=True)
base

Age Nationality  Overall  Potential                 Club    Value  \
0       31   Argentina       94         94         FC Barcelona  €110.5M   
1       33    Portugal       94         94             Juventus     €77M   
2       26      Brazil       92         93  Paris Saint-Germain  €118.5M   
3       27       Spain       91         93    Manchester United     €72M   
4       27     Belgium       91         92      Manchester City    €102M   
...    ...         ...      ...        ...                  ...      ...   
18202   19     England       47         65      Crewe Alexandra     €60K   
18203   19      Sweden       47         63       Trelleborgs FF     €60K   
18204   16     England       47         67     Cambridge United     €60K   
18205   17     England       47         66      Tranmere Rovers     €60K   
18206   16     England       46         66      Tranmere Rovers     €60K   

        Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0      €565K     2202           Left                       5.0        4.0   
1      €405K     2228          Right                       5.0        4.0   
2      €290K     2143          Right                       5.0        5.0   
3      €260K     1471          Right                       4.0        3.0   
4      €355K     2281          Right                       4.0        5.0   
...      ...      ...            ...                       ...        ...   
18202    €1K     1307          Right                       1.0        2.0   
18203    €1K     1098          Right                       1.0        2.0   
18204    €1K     1189          Right                       1.0        3.0   
18205    €1K     1228          Right                       1.0        3.0   
18206    €1K     1321          Right                       1.0        3.0   

       Skill Moves       Work Rate   Body Type Position        Joined  \
0              4.0  Medium/ Medium       Messi       RF   Jul 1, 2004   
1              5.0       High/ Low  C. Ronaldo       ST  Jul 10, 2018   
2              5.0    High/ Medium      Neymar       LW   Aug 3, 2017   
3              1.0  Medium/ Medium        Lean       GK   Jul 1, 2011   
4              4.0      High/ High      Normal      RCM  Aug 30, 2015   
...            ...             ...         ...      ...           ...   
18202          2.0  Medium/ Medium        Lean       CM   May 3, 2017   
18203          2.0  Medium/ Medium      Normal       ST  Mar 19, 2018   
18204          2.0  Medium/ Medium      Normal       ST   Jul 1, 2017   
18205          2.0  Medium/ Medium        Lean       RW  Apr 24, 2018   
18206          2.0  Medium/ Medium        Lean       CM  Oct 30, 2018   

      Loaned From Contract Valid Until Height  Weight    LS    ST    RS    LW  \
0             NaN                 2021    5'7  159lbs  88+2  88+2  88+2  92+2   
1             NaN                 2022    6'2  183lbs  91+3  91+3  91+3  89+3   
2             NaN                 2022    5'9  150lbs  84+3  84+3  84+3  89+3   
3             NaN                 2020    6'4  168lbs   NaN   NaN   NaN   NaN   
4             NaN                 2023   5'11  154lbs  82+3  82+3  82+3  87+3   
...           ...                  ...    ...     ...   ...   ...   ...   ...   
18202         NaN                 2019    5'9  134lbs  42+2  42+2  42+2  44+2   
18203         NaN                 2020    6'3  170lbs  45+2  45+2  45+2  39+2   
18204         NaN                 2021    5'8  148lbs  45+2  45+2  45+2  45+2   
18205         NaN                 2019   5'10  154lbs  47+2  47+2  47+2  47+2   
18206         NaN                 2019   5'10  176lbs  43+2  43+2  43+2  45+2   

         LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0      93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1      90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2      89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+

Por otro lado, las características Joined and Loaned From son más importantes que las anteriores a la hora de construir el modelo de predicción, pero no lo suficiente para contar con ellas. Por lo tanto, las eliminaremos

In [9]:
base.drop(['Joined','Loaned From','Contract Valid Until'], axis = 1,inplace=True)
base.head()

Age Nationality  Overall  Potential                 Club    Value   Wage  \
0   31   Argentina       94         94         FC Barcelona  €110.5M  €565K   
1   33    Portugal       94         94             Juventus     €77M  €405K   
2   26      Brazil       92         93  Paris Saint-Germain  €118.5M  €290K   
3   27       Spain       91         93    Manchester United     €72M  €260K   
4   27     Belgium       91         92      Manchester City    €102M  €355K   

   Special Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
0     2202           Left                       5.0        4.0          4.0   
1     2228          Right                       5.0        4.0          5.0   
2     2143          Right                       5.0        5.0          5.0   
3     1471          Right                       4.0        3.0          1.0   
4     2281          Right                       4.0        5.0          4.0   

        Work Rate   Body Type Position Height  Weight    LS    ST    RS    LW  \
0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2  88+2  88+2  92+2   
1       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3  91+3  91+3  89+3   
2    High/ Medium      Neymar       LW    5'9  150lbs  84+3  84+3  84+3  89+3   
3  Medium/ Medium        Lean       GK    6'4  168lbs   NaN   NaN   NaN   NaN   
4      High/ High      Normal      RCM   5'11  154lbs  82+3  82+3  82+3  87+3   

     LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM  \
0  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2   
1  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3   
2  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3  87+3  88+3   

    LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  Crossing  \
0  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2      84.0   
1  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3      84.0   
2  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3      79.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN      17.0   
4  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3      93.0   

   Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  \
0       95.0             70.0          90.0     86.0       97.0   93.0   
1       94.0             89.0          81.0     87.0       88.0   81.0   
2       87.0             62.0          84.0     84.0       96.0   88.0   
3       13.0             21.0          50.0     13.0       18.0   21.0   
4       82.0             55.0          92.0     82.0       86.0   85.0   

   FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  Agility  \
0        94.0         87.0         96.0          91.0         86.0     91.0   
1        76.0         77.0         94.0          89.0         91.0     87.0   
2        87.0         78.0         95.0          94.0         90.0     96.0   
3        19.0         51.0         42.0          57.0         58.0     60.0   
4        83.0         91.0         91.0          78.0         76.0     79.0   

   Reactions  Balance  ShotPower  Jumping  Stamina  Strength  LongShots  \
0       95.0     95.0       85.0     68.0     72.0      59.0       94.0   
1       96.0     70.0       95.0     95.0     88.0      79.0       93.0   
2       94.0     84.0       80.0     61.0     81.0      49.0       82.0   
3       90.0     43.0       31.0     67.0     43.0      64.0       12.0   
4       91.0     77.0       91.0     63.0     90.0      75.0       91.0   

   Aggression  Interceptions  Positioning  Vision  Penalties  Composure  \
0        48.0           22.0         94.0    94.0       75.0       96.0   
1        63.0           29.0         95.0    82.0       85.0       95.0   
2        56.0           36.0         89.0    87.0       

## Gestión de valores no númericos

La mayoría de datos no están en el formato correcto, en float y contienen NaNs por ello, habrá que tratarlos.

Observamos que hay valores que no están en el formato correcto, entre ellos : Valor, Salario, Cláusula de recisión,  Peso, Altura

El valor, salario y cláusula de recisión están en formato monetario, en € concretamente. Por otro lado,  las abreviaturas 'K' y 'M' representan mil euros y millón de euros respectivamente.

Esta función nos transforma value a float. 

In [10]:
def value_to_float(x):
    """
    From K and M to float.
    
    """
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

Esta función nos transforma wage a float.

In [11]:
def wage_to_float(x):
    """
    From K and M to float.
    
    """
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

Esta función nos transforma release clause a float.

In [12]:
def releaseclause_to_float(x):
    """
    From K and M to float.
    
    """
    x = x.replace('€', '')
    ret_val = 0.0
    
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000.0
    return ret_val

Las funciones funcionaron bien en las columnas Valor y Salario,como se puede comprobar abajo.Por otro lado, en la clausula de recisión me dió error pues contiene valores NaN.

In [13]:
base["Value"] = base["Value"].apply(value_to_float)
base["Wage"] = base["Wage"].apply(wage_to_float)
base.head()

Age Nationality  Overall  Potential                 Club        Value  \
0   31   Argentina       94         94         FC Barcelona  110500000.0   
1   33    Portugal       94         94             Juventus   77000000.0   
2   26      Brazil       92         93  Paris Saint-Germain  118500000.0   
3   27       Spain       91         93    Manchester United   72000000.0   
4   27     Belgium       91         92      Manchester City  102000000.0   

       Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  565000.0     2202           Left                       5.0        4.0   
1  405000.0     2228          Right                       5.0        4.0   
2  290000.0     2143          Right                       5.0        5.0   
3  260000.0     1471          Right                       4.0        3.0   
4  355000.0     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Position Height  Weight    LS  \
0          4.0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2   
1          5.0       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3   
2          5.0    High/ Medium      Neymar       LW    5'9  150lbs  84+3   
3          1.0  Medium/ Medium        Lean       GK    6'4  168lbs   NaN   
4          4.0      High/ High      Normal      RCM   5'11  154lbs  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.

Se puede comprobar con esta función.

In [14]:
base.loc[base['Release Clause'].isnull()]

Age Nationality  Overall  Potential                            Club  \
28      26    Colombia       88         89               FC Bayern München   
38      30   Argentina       88         88                           Milan   
91      29      Brazil       85         85  Guangzhou Evergrande Taobao FC   
166     24      Brazil       83         90  Guangzhou Evergrande Taobao FC   
176     24     Croatia       83         89                         Chelsea   
...    ...         ...      ...        ...                             ...   
17979   21    China PR       51         60             Guizhou Hengfeng FC   
18026   21    China PR       50         59             Guizhou Hengfeng FC   
18031   20    China PR       50         61                  Stabæk Fotball   
18056   19       Italy       50         65                          Ascoli   
18183   44     England       48         48                Cambridge United   

            Value      Wage  Special Preferred Foot  International Reputation  \
28     69500000.0  315000.0     2171           Left                       4.0   
38     57000000.0  245000.0     1965          Right                       4.0   
91     37000000.0  235000.0     2245          Right                       3.0   
166    36500000.0   18000.0     2142           Left                       2.0   
176    35000000.0  165000.0     2115          Right                       3.0   
...           ...       ...      ...            ...                       ...   
17979     60000.0    2000.0     1459          Right                       1.0   
18026     50000.0    2000.0     1440          Right                       1.0   
18031     40000.0    2000.0     1278          Right                       1.0   
18056     60000.0    3000.0     1450           Left                       1.0   
18183         0.0    1000.0      774          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate Body Type Position Height  \
28           3.0          4.0  Medium/ Medium    Normal      LAM   5'11   
38           4.0          3.0    High/ Medium    Normal       LS    6'1   
91           3.0          3.0      High/ High      Lean      LDM    6'0   
166          4.0          4.0    High/ Medium    Normal      CAM    6'3   
176          4.0          4.0      High/ High    Normal      LCM    5'9   
...          ...          ...             ...       ...      ...    ...   
17979        2.0          2.0  Medium/ Medium    Normal       CM    6'0   
18026        2.0          2.0  Medium/ Medium      Lean       LM   5'10   
18031        3.0          2.0  Medium/ Medium    Normal       RB    5'7   
18056        3.0          2.0  Medium/ Medium      Lean       CM   5'10   
18183        2.0          1.0  Medium/ Medium    Normal       GK    6'2   

       Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
28     172lbs  80+3  80+3  80+3  84+3  83+3  83+3  83+3  84+3  85+3  85+3   
38     196lbs  85+3  85+3  85+3  81+3  83+3  83+3  83+3  81+3  80+3  80+3   
91     157lbs  81+3  81+3  81+3  79+3  81+3  81+3  81+3  79+3  81+3  81+3   
166    176lbs  80+2  80+2  80+2  81+2  81+2  81+2  81+2  81+2  81+2  81+2   
176    172lbs  71+3  71+3  71+3  79+3  78+3  78+3  78+3  79+3  81+3  81+3   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17979  172lbs  49+2  49+2  49+2  50+2  50+2  50+2  50+2  50+2  50+2  50+2   
18026  143lbs  44+2  44+2  44+2  47+2  46+2  46+2  46+2  47+2  48+2  48+2   
18031  143lbs  37+2  37+2  37+2  42+2  40+2  40+2  40+2  42+2  39+2  39+2   
18056  143lbs  44+2  44+2  44+2  47+2  45+2  45+2  45+2  47+2  47+2  47+2   
18183  190lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
28     85+3  83+3  81+3  81+3  81+3  83+3  69+3  68+3  68+3  68+3  69+3  65+3   
38     80+3  78+3  71+3  71+3  71+3  78+3  55+3  53+3  53+3  53+3  55+3  51+3   
91     81+3  79+3  82+3  82+

Como gestioné esos NaNs? Al haber una gran cantidad de jugadores con esta característica y sobretodo por ser relativa al valor, no era una opción viable eliminarlos.Por ello, para resolverlo, sustituí(imputé) los NaNs por 0

In [15]:
base["Release Clause"].fillna('0€', inplace=True)

Comprobamos como la función funciona correctamente, pues en estos jugadores la característica Release Clause es igual a 0 €

In [16]:
base.loc[base['Release Clause'] == '0€']

Age Nationality  Overall  Potential                            Club  \
28      26    Colombia       88         89               FC Bayern München   
38      30   Argentina       88         88                           Milan   
91      29      Brazil       85         85  Guangzhou Evergrande Taobao FC   
166     24      Brazil       83         90  Guangzhou Evergrande Taobao FC   
176     24     Croatia       83         89                         Chelsea   
...    ...         ...      ...        ...                             ...   
17979   21    China PR       51         60             Guizhou Hengfeng FC   
18026   21    China PR       50         59             Guizhou Hengfeng FC   
18031   20    China PR       50         61                  Stabæk Fotball   
18056   19       Italy       50         65                          Ascoli   
18183   44     England       48         48                Cambridge United   

            Value      Wage  Special Preferred Foot  International Reputation  \
28     69500000.0  315000.0     2171           Left                       4.0   
38     57000000.0  245000.0     1965          Right                       4.0   
91     37000000.0  235000.0     2245          Right                       3.0   
166    36500000.0   18000.0     2142           Left                       2.0   
176    35000000.0  165000.0     2115          Right                       3.0   
...           ...       ...      ...            ...                       ...   
17979     60000.0    2000.0     1459          Right                       1.0   
18026     50000.0    2000.0     1440          Right                       1.0   
18031     40000.0    2000.0     1278          Right                       1.0   
18056     60000.0    3000.0     1450           Left                       1.0   
18183         0.0    1000.0      774          Right                       1.0   

       Weak Foot  Skill Moves       Work Rate Body Type Position Height  \
28           3.0          4.0  Medium/ Medium    Normal      LAM   5'11   
38           4.0          3.0    High/ Medium    Normal       LS    6'1   
91           3.0          3.0      High/ High      Lean      LDM    6'0   
166          4.0          4.0    High/ Medium    Normal      CAM    6'3   
176          4.0          4.0      High/ High    Normal      LCM    5'9   
...          ...          ...             ...       ...      ...    ...   
17979        2.0          2.0  Medium/ Medium    Normal       CM    6'0   
18026        2.0          2.0  Medium/ Medium      Lean       LM   5'10   
18031        3.0          2.0  Medium/ Medium    Normal       RB    5'7   
18056        3.0          2.0  Medium/ Medium      Lean       CM   5'10   
18183        2.0          1.0  Medium/ Medium    Normal       GK    6'2   

       Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
28     172lbs  80+3  80+3  80+3  84+3  83+3  83+3  83+3  84+3  85+3  85+3   
38     196lbs  85+3  85+3  85+3  81+3  83+3  83+3  83+3  81+3  80+3  80+3   
91     157lbs  81+3  81+3  81+3  79+3  81+3  81+3  81+3  79+3  81+3  81+3   
166    176lbs  80+2  80+2  80+2  81+2  81+2  81+2  81+2  81+2  81+2  81+2   
176    172lbs  71+3  71+3  71+3  79+3  78+3  78+3  78+3  79+3  81+3  81+3   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17979  172lbs  49+2  49+2  49+2  50+2  50+2  50+2  50+2  50+2  50+2  50+2   
18026  143lbs  44+2  44+2  44+2  47+2  46+2  46+2  46+2  47+2  48+2  48+2   
18031  143lbs  37+2  37+2  37+2  42+2  40+2  40+2  40+2  42+2  39+2  39+2   
18056  143lbs  44+2  44+2  44+2  47+2  45+2  45+2  45+2  47+2  47+2  47+2   
18183  190lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
28     85+3  83+3  81+3  81+3  81+3  83+3  69+3  68+3  68+3  68+3  69+3  65+3   
38     80+3  78+3  71+3  71+3  71+3  78+3  55+3  53+3  53+3  53+3  55+3  51+3   
91     81+3  79+3  82+3  82+

Ahora si podemos aplicar la función que transforma los valores monetarios de la cláusula de recisión a float

In [17]:
base["Release Clause"] = base["Release Clause"].apply(wage_to_float)

Comprobamos que en la última columna la característica Cláusula de reicisión es float

In [18]:
base.head()

Age Nationality  Overall  Potential                 Club        Value  \
0   31   Argentina       94         94         FC Barcelona  110500000.0   
1   33    Portugal       94         94             Juventus   77000000.0   
2   26      Brazil       92         93  Paris Saint-Germain  118500000.0   
3   27       Spain       91         93    Manchester United   72000000.0   
4   27     Belgium       91         92      Manchester City  102000000.0   

       Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  565000.0     2202           Left                       5.0        4.0   
1  405000.0     2228          Right                       5.0        4.0   
2  290000.0     2143          Right                       5.0        5.0   
3  260000.0     1471          Right                       4.0        3.0   
4  355000.0     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Position Height  Weight    LS  \
0          4.0  Medium/ Medium       Messi       RF    5'7  159lbs  88+2   
1          5.0       High/ Low  C. Ronaldo       ST    6'2  183lbs  91+3   
2          5.0    High/ Medium      Neymar       LW    5'9  150lbs  84+3   
3          1.0  Medium/ Medium        Lean       GK    6'4  168lbs   NaN   
4          4.0      High/ High      Normal      RCM   5'11  154lbs  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.

Ahora me fijo en otras características que tampoco están en el formato adecuado como el peso y la altura que pueden ser importantes a la hora de predecir el valor del jugador. Por ello, las transformaremos a float con estas funciones.

In [19]:
def height_to_float(x):
    """
    x'x -> x.x
    """
    x = x.split("'")
    
    if(len(x[1])==1): x = float(x[0]) + float(x[1])/10
    else:
        y = float(x[1])
        y = y/100
        x = float(x[0])+y

    return x

In [20]:
base.loc[base['Height'].isnull()]

Age       Nationality  Overall  Potential  \
13236   33          Scotland       62         62   
13237   29         Nicaragua       62         62   
13238   35           England       62         62   
13239   20             Italy       62         72   
13240   24           England       62         66   
13241   21            Poland       62         72   
13242   23         Argentina       62         70   
13243   19           England       62         78   
13244   30           Denmark       62         62   
13245   21            Brazil       62         69   
13246   22            France       62         68   
13247   28           Austria       62         62   
13248   24  Northern Ireland       62         69   
13249   27          China PR       62         62   
13250   29          Paraguay       62         62   
13251   34           Belgium       62         62   
13252   22           Albania       62         70   
13253   31          Scotland       62         62   
13254   17           Germany       62         82   
13255   25           Germany       62         65   
13256   26           England       62         62   
13257   25    Korea Republic       62         65   
13258   23      Saudi Arabia       62         67   
13259   27             Chile       62         65   
13260   20             Spain       62         69   
13261   25      Saudi Arabia       62         64   
13262   24           Germany       62         66   
13263   25           Austria       62         66   
13264   29            France       62         62   
13265   28           England       62         62   
13266   20            Sweden       62         73   
13267   22          Colombia       62         70   
13268   37            Poland       62         62   
13269   26          Colombia       62         64   
13270   19            Brazil       62         77   
13271   26             Chile       62         63   
13272   22          Scotland       62         70   
13273   25           Estonia       62         67   
13274   27             Italy       62         62   
13275   23         Venezuela       62         70   
13276   26      Saudi Arabia       62         63   
13277   26             Chile       62         62   
13278   17       Netherlands       62         79   
13279   22             Italy       62         69   
13280   19            France       62         77   
13281   27    Korea Republic       62         62   
13282   25            Mexico       62         65   
13283   25          China PR       62         66   

                                Club     Value    Wage  Special  \
13236                       Rochdale  120000.0  1000.0     1510   
13237                Boyacá Chicó FC  300000.0  1000.0     1532   
13238                   Notts County  140000.0  3000.0     1573   
13239                        Brescia  425000.0  1000.0     1610   
13240         Hamilton Academical FC  400000.0  1000.0     1481   
13241                  Śląsk Wrocław  425000.0  1000.0     1692   
13242         Club Atlético Aldosivi  450000.0  2000.0     1663   
13243                        Everton  600000.0  5000.0     1328   
13244                       Hobro IK  230000.0  2000.0     1244   
13245                   HJK Helsinki  425000.0  1000.0     1549   
13246                     AS Béziers  425000.0  2000.0     1494   
13247                 SV Mattersburg  240000.0  3000.0     1630   
13248                Tranmere Rovers  375000.0  2000.0     1461   
13249  Shanghai Greenland Shenhua FC  250000.0  3000.0     1636   
13250               Itagüí Leones FC  300000.0  1000.0     1454   
13251                      NAC Breda  150000.0  2000.0     1665   
13252                       Malmö FF  375000.0  1000.0     1587   
13253                Carlisle United  200000.0  2000.0     1535   
13254                  VfB Stuttgart  550000.0  2000.0     1418   
13255         Hamilton Academical FC  325000.0  1000.0     1693   
13256                      Dundee FC  325000.0  100

Observamos que esta lista escasa de jugadores contiene excesivos NaN, por lo tanto, es mejor eliminarlos directamente ya que nos darán también problemas con la altura . Además van a perjudicar a nuestro dataset de entrenamiento pues no contienen información.

In [21]:
base = base.dropna(subset = ['Height'])

In [22]:
base["Height"] =base["Height"].apply(height_to_float)
base.head()

C:\phyton\envs\analisi_dades\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Age Nationality  Overall  Potential                 Club        Value  \
0   31   Argentina       94         94         FC Barcelona  110500000.0   
1   33    Portugal       94         94             Juventus   77000000.0   
2   26      Brazil       92         93  Paris Saint-Germain  118500000.0   
3   27       Spain       91         93    Manchester United   72000000.0   
4   27     Belgium       91         92      Manchester City  102000000.0   

       Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  565000.0     2202           Left                       5.0        4.0   
1  405000.0     2228          Right                       5.0        4.0   
2  290000.0     2143          Right                       5.0        5.0   
3  260000.0     1471          Right                       4.0        3.0   
4  355000.0     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Position  Height  Weight    LS  \
0          4.0  Medium/ Medium       Messi       RF    5.70  159lbs  88+2   
1          5.0       High/ Low  C. Ronaldo       ST    6.20  183lbs  91+3   
2          5.0    High/ Medium      Neymar       LW    5.90  150lbs  84+3   
3          1.0  Medium/ Medium        Lean       GK    6.40  168lbs   NaN   
4          4.0      High/ High      Normal      RCM    5.11  154lbs  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0 

In [23]:
def weight_to_float(x):
    """
    From lbs to float.
    
    """
    if 'lbs' in x:
            ret_val = float(x.replace('lbs', ' '))
    return ret_val

In [24]:
base["Weight"] = base["Weight"].apply(weight_to_float)
base.head()

C:\phyton\envs\analisi_dades\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Age Nationality  Overall  Potential                 Club        Value  \
0   31   Argentina       94         94         FC Barcelona  110500000.0   
1   33    Portugal       94         94             Juventus   77000000.0   
2   26      Brazil       92         93  Paris Saint-Germain  118500000.0   
3   27       Spain       91         93    Manchester United   72000000.0   
4   27     Belgium       91         92      Manchester City  102000000.0   

       Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  565000.0     2202           Left                       5.0        4.0   
1  405000.0     2228          Right                       5.0        4.0   
2  290000.0     2143          Right                       5.0        5.0   
3  260000.0     1471          Right                       4.0        3.0   
4  355000.0     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Position  Height  Weight    LS  \
0          4.0  Medium/ Medium       Messi       RF    5.70   159.0  88+2   
1          5.0       High/ Low  C. Ronaldo       ST    6.20   183.0  91+3   
2          5.0    High/ Medium      Neymar       LW    5.90   150.0  84+3   
3          1.0  Medium/ Medium        Lean       GK    6.40   168.0   NaN   
4          4.0      High/ High      Normal      RCM    5.11   154.0  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0 

Ahora, comprobamos que existen valores como Nacionalidad,Club,Tipo de Cuerpo, Posición y Pie Preferente que no tienen valores númericos y son valores categóricos. Para ello, aplicaremos la técnica(hot-encoding) usada basándome en la del enunciado con Club y convertiremos los valores a binarios. De esta manera, nos evitaremos problemas con NaNs y podremos convertir los valores no númericos a númericos

In [25]:
wf = base.pop("Nationality")
base = pd.concat([base.reset_index(drop=True), pd.get_dummies(wf, prefix='nat').reset_index(drop=True)],axis=1, sort=False)
filtro_Pos = [col for col in base if col.startswith('nat')]
base.drop(columns=filtro_Pos)

wf = base.pop("Club")
base = pd.concat([base.reset_index(drop=True), pd.get_dummies(wf, prefix='clb').reset_index(drop=True)],axis=1, sort=False)
filtro_Pos = [col for col in base if col.startswith('clb')]
base.drop(columns=filtro_Pos)

wf = base.pop("Body Type")
base = pd.concat([base.reset_index(drop=True), pd.get_dummies(wf, prefix='bdy').reset_index(drop=True)],axis=1, sort=False)
filtro_Pos = [col for col in base if col.startswith('bdy')]
base.drop(columns=filtro_Pos)

wf = base.pop("Position")
base = pd.concat([base.reset_index(drop=True), pd.get_dummies(wf, prefix='pos').reset_index(drop=True)],axis=1, sort=False)
filtro_Pos = [col for col in base if col.startswith('pos')]
base.drop(columns=filtro_Pos)

wf = base.pop("Preferred Foot")
base = pd.concat([base.reset_index(drop=True), pd.get_dummies(wf, prefix='foot').reset_index(drop=True)],axis=1, sort=False)
filtro_Pos = [col for col in base if col.startswith('foot')]
base.drop(columns=filtro_Pos)

base.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves       Work Rate  Height  \
0                       5.0        4.0          4.0  Medium/ Medium    5.70   
1                       5.0        4.0          5.0       High/ Low    6.20   
2                       5.0        5.0          5.0    High/ Medium    5.90   
3                       4.0        3.0          1.0  Medium/ Medium    6.40   
4                       4.0        5.0          4.0      High/ High    5.11   

   Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM  \
0   159.0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2   
1   183.0  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3   
2   150.0  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3   
3   168.0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4   154.0  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3   

     LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB  \
0  91+2  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2   
1  88+3  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3   
2  88+3  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  88+3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3   

     CB   RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  \
0  47+2  47+2  59+2      84.0       95.0             70.0          90.0   
1  53+3  53+3  61+3      84.0       94.0             89.0          81.0   
2  47+3  47+3  60+3      79.0       87.0             62.0          84.0   
3   NaN   NaN   NaN      17.0       13.0             21.0          50.0   
4  66+3  66+3  73+3      93.0       82.0             55.0          92.0   

   Volleys  Dribbling  Curve  FKAccuracy  LongPassing  BallControl  \
0     86.0       97.0   93.0        94.0         87.0         96.0   
1     87.0       88.0   81.0        76.0         77.0         94.0   
2     84.0       96.0   88.0        87.0         78.0         95.0   
3     13.0       18.0   21.0        19.0         51.0         42.0   
4     82.0       86.0   85.0        83.0         91.0         91.0   

   Acceleration  SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  \
0          91.0         86.0     91.0       95.0     95.0       85.0     68.0   
1          89.0         91.0     87.0       96.0     70.0       95.0     95.0   
2          94.0         90.0     96.0       94.0     84.0       80.0     61.0   
3          57.0         58.0     60.0       90.0     43.0       31.0     67.0   
4          78.0         76.0     79.0       91.0     77.0       91.0     63.0   

   Stamina  Strength  LongShots  Aggression  Interceptions  Positioning  \
0     72.0      59.0       94.0        48.0           22.0         94.0   
1     88.0      79.0       93.0        63.0           29.0         95.0   
2     81.0      49.0       82.0        56.0           36.0         89.0   
3     43.0      64.0       12.0        38.0           30.0         12.0   
4     90.0      75.0       91.0        76.0           61.0         87.0   

   Vision  Penalties  Composure  Marking  StandingTackle  SlidingTackle  \
0    94.0       75.0       96.0     33.0            28.0           26.0   
1    82.0       85.0       95.0     28.0            31.0           23.0   
2    87.0       81.0       94.0     27.0            24.0           33.0   
3    68.0       40.0       68.0     15.0            21.0           13.0   
4    94.0       79.0       88.0 

Otros valor no númerico es Work Rate, este se divide según su valoración en ataque y defensa, el valor antes de la barra es el de ataque y el valor de despues de la barra es el de defensa. Sigue este formato --> "Valoración en ataque"/ "Valoracion en Defensa".Este valor puede ser "High","Medium" o "Low".  
Para ello, a través de esta función, dividiremos en dos columnas el valor en ataque y el valor en defensa, de esta manera tendremos más facilidad de tratarlo.

In [26]:
base[['Attack Work Rate','Defense Work Rate']] = base['Work Rate'].str.split('/',n=1,expand= True)
base.pop("Work Rate")
base.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight    LS  \
0                       5.0        4.0          4.0    5.70   159.0  88+2   
1                       5.0        4.0          5.0    6.20   183.0  91+3   
2                       5.0        5.0          5.0    5.90   150.0  84+3   
3                       4.0        3.0          1.0    6.40   168.0   NaN   
4                       4.0        5.0          4.0    5.11   154.0  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0           13.0      90.0   
4       79.0       88.0     68.0            58.0           51.0      1

Y ahora volveremos a aplicar la técnica utilizada anteriormente para convertir valores categóricos a binarios

In [27]:
wf = base.pop("Attack Work Rate")
base = pd.concat([base.reset_index(drop=True), pd.get_dummies(wf, prefix='AWR').reset_index(drop=True)],axis=1, sort=False)
filtro_Pos = [col for col in base if col.startswith('AWR')]
base.drop(columns=filtro_Pos)

wf = base.pop("Defense Work Rate")
base = pd.concat([base.reset_index(drop=True), pd.get_dummies(wf, prefix='DWR').reset_index(drop=True)],axis=1, sort=False)
filtro_Pos = [col for col in base if col.startswith('DWR')]
base.drop(columns=filtro_Pos)
base.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight    LS  \
0                       5.0        4.0          4.0    5.70   159.0  88+2   
1                       5.0        4.0          5.0    6.20   183.0  91+3   
2                       5.0        5.0          5.0    5.90   150.0  84+3   
3                       4.0        3.0          1.0    6.40   168.0   NaN   
4                       4.0        5.0          4.0    5.11   154.0  82+3   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2   
1  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3   
2  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2   
1  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3   
2  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  59+2      84.0       95.0             70.0          90.0     86.0   
1  61+3      84.0       94.0             89.0          81.0     87.0   
2  60+3      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  73+3      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0           13.0      90.0   
4       79.0       88.0     68.0            58.0           51.0      1

Observamos que deberíamos tratar los valores de rendimiento del jugador en cada posición(LS	ST	RS	LW	LF	CF	RF	RW	LAM	CAM	RAM	LM	LCM	CM	RCM	RM	LWB	LDM	CDM	RDM	RWB	LB	LCB	CB	RCB	RB) . Para ello, simplemte sumamos los valores para que quede en un solo valor númerico.  Con esta función se sumarán los dos valores: 

In [28]:
def add_pos(x):    
    """
    Suma de los valores. 
    
    """
    
    if(x != x):
        return np.nan
    
    values = x.split('+')
    
    return (float(values[0])+float(values[1]))

Utilizamos dicha función para aplicarla a cada posición posible

In [29]:
base["LS"] = base["LS"].apply(add_pos)
base["ST"] = base["ST"].apply(add_pos)
base["RS"] = base["RS"].apply(add_pos)
base["LW"] = base["LW"].apply(add_pos)
base["LF"] = base["LF"].apply(add_pos)
base["CF"] = base["CF"].apply(add_pos)
base["RF"] = base["RF"].apply(add_pos)
base["RW"] = base["RW"].apply(add_pos)
base["LAM"] = base["LAM"].apply(add_pos)
base["CAM"] = base["CAM"].apply(add_pos)
base["RAM"] = base["RAM"].apply(add_pos)
base["LM"] = base["LM"].apply(add_pos)
base["LCM"] = base["LCM"].apply(add_pos)
base["CM"] = base["CM"].apply(add_pos)
base["RCM"] = base["RCM"].apply(add_pos)
base["RM"] = base["RM"].apply(add_pos)
base["LWB"] = base["LWB"].apply(add_pos)
base["LDM"] = base["LDM"].apply(add_pos)
base["CDM"] = base["CDM"].apply(add_pos)
base["RDM"] = base["RDM"].apply(add_pos)
base["RWB"] = base["RWB"].apply(add_pos)
base["LB"] = base["LB"].apply(add_pos)
base["LCB"] = base["LCB"].apply(add_pos)
base["CB"] = base["CB"].apply(add_pos)
base["RCB"] = base["RCB"].apply(add_pos)
base["RB"] = base["RB"].apply(add_pos)
base.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight    LS  \
0                       5.0        4.0          4.0    5.70   159.0  90.0   
1                       5.0        4.0          5.0    6.20   183.0  94.0   
2                       5.0        5.0          5.0    5.90   150.0  87.0   
3                       4.0        3.0          1.0    6.40   168.0   NaN   
4                       4.0        5.0          4.0    5.11   154.0  85.0   

     ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM  \
0  90.0  90.0  94.0  95.0  95.0  95.0  94.0  95.0  95.0  95.0  93.0  86.0   
1  94.0  94.0  92.0  93.0  93.0  93.0  92.0  91.0  91.0  91.0  91.0  84.0   
2  87.0  87.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  92.0  91.0  84.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  85.0  85.0  90.0  90.0  90.0  90.0  90.0  91.0  91.0  91.0  91.0  90.0   

     CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB  \
0  86.0  86.0  93.0  66.0  63.0  63.0  63.0  66.0  61.0  49.0  49.0  49.0   
1  84.0  84.0  91.0  68.0  64.0  64.0  64.0  68.0  64.0  56.0  56.0  56.0   
2  84.0  84.0  91.0  68.0  63.0  63.0  63.0  68.0  63.0  50.0  50.0  50.0   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  90.0  90.0  91.0  80.0  80.0  80.0  80.0  80.0  76.0  69.0  69.0  69.0   

     RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  61.0      84.0       95.0             70.0          90.0     86.0   
1  64.0      84.0       94.0             89.0          81.0     87.0   
2  63.0      79.0       87.0             62.0          84.0     84.0   
3   NaN      17.0       13.0             21.0          50.0     13.0   
4  76.0      93.0       82.0             55.0          92.0     82.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       18.0   21.0        19.0         51.0         42.0          57.0   
4       86.0   85.0        83.0         91.0         91.0          78.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         58.0     60.0       90.0     43.0       31.0     67.0     43.0   
4         76.0     79.0       91.0     77.0       91.0     63.0     90.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      64.0       12.0        38.0           30.0         12.0    68.0   
4      75.0       91.0        76.0           61.0         87.0    94.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.0       9.0   
3       40.0       68.0     15.0            21.0           13.0      90.0   
4       79.0       88.0     68.0            58.0           51.0      1

Por último, vemos como David de Gea (y el resto de porteros) tiene NaNs en estos valores, para ello podríamos sustuir estos modificadores de posición por ceros, pero no me parece la solución más óptima. Una posible solución, es rellenar las NaNs con las medias de las columnas, no será la solución más eficiente pero es una posible solución alterna a tener que imputar ceros.

In [30]:
base.fillna(base.mean(), inplace=True)
base.head()

Age  Overall  Potential        Value      Wage  Special  \
0   31       94         94  110500000.0  565000.0     2202   
1   33       94         94   77000000.0  405000.0     2228   
2   26       92         93  118500000.0  290000.0     2143   
3   27       91         93   72000000.0  260000.0     1471   
4   27       91         92  102000000.0  355000.0     2281   

   International Reputation  Weak Foot  Skill Moves  Height  Weight      LS  \
0                       5.0        4.0          4.0    5.70   159.0  90.000   
1                       5.0        4.0          5.0    6.20   183.0  94.000   
2                       5.0        5.0          5.0    5.90   150.0  87.000   
3                       4.0        3.0          1.0    6.40   168.0  59.836   
4                       4.0        5.0          4.0    5.11   154.0  85.000   

       ST      RS         LW         LF         CF         RF         RW  \
0  90.000  90.000  94.000000  95.000000  95.000000  95.000000  94.000000   
1  94.000  94.000  92.000000  93.000000  93.000000  93.000000  92.000000   
2  87.000  87.000  92.000000  92.000000  92.000000  92.000000  92.000000   
3  59.836  59.836  61.058181  60.739921  60.739921  60.739921  61.058181   
4  85.000  85.000  90.000000  90.000000  90.000000  90.000000  90.000000   

        LAM       CAM       RAM         LM        LCM         CM        RCM  \
0  95.00000  95.00000  95.00000  93.000000  86.000000  86.000000  86.000000   
1  91.00000  91.00000  91.00000  91.000000  84.000000  84.000000  84.000000   
2  92.00000  92.00000  92.00000  91.000000  84.000000  84.000000  84.000000   
3  61.00366  61.00366  61.00366  61.755241  60.241409  60.241409  60.241409   
4  91.00000  91.00000  91.00000  91.000000  90.000000  90.000000  90.000000   

          RM        LWB        LDM        CDM        RDM        RWB  \
0  93.000000  66.000000  63.000000  63.000000  63.000000  66.000000   
1  91.000000  68.000000  64.000000  64.000000  64.000000  68.000000   
2  91.000000  68.000000  63.000000  63.000000  63.000000  68.000000   
3  61.755241  59.533805  58.845677  58.845677  58.845677  59.533805   
4  91.000000  80.000000  80.000000  80.000000  80.000000  80.000000   

          LB        LCB         CB        RCB         RB  Crossing  Finishing  \
0  61.000000  49.000000  49.000000  49.000000  61.000000      84.0       95.0   
1  64.000000  56.000000  56.000000  56.000000  64.000000      84.0       94.0   
2  63.000000  50.000000  50.000000  50.000000  63.000000      79.0       87.0   
3  58.827317  57.635157  57.635157  57.635157  58.827317      17.0       13.0   
4  76.000000  69.000000  69.000000  69.000000  76.000000      93.0       82.0   

   HeadingAccuracy  ShortPassing  Volleys  Dribbling  Curve  FKAccuracy  \
0             70.0          90.0     86.0       97.0   93.0        94.0   
1             89.0          81.0     87.0       88.0   81.0        76.0   
2             62.0          84.0     84.0       96.0   88.0        87.0   
3             21.0          50.0     13.0       18.0   21.0        19.0   
4             55.0          92.0     82.0       86.0   85.0        83.0   

   LongPassing  BallControl  Acceleration  SprintSpeed  Agility  Reactions  \
0         87.0         96.0          91.0         86.0     91.0       95.0   
1         77.0         94.0          89.0         91.0     87.0       96.0   
2         78.0         95.0          94.0         90.0     96.0       94.0   
3         51.0         42.0          57.0         58.0     60.0       90.0   
4         91.0         91.0          78.0         76.0     79.0       91.0   

   Balance  ShotPower  Jumping  Stamina  Strength  LongShots  Aggression  \
0     95.0       85.0     68.0     72.0      59.0       94.0        48.0   
1     70.0       95.0     95.0     88.0      79.0       93.0        63.0   
2     84.0       80.0     61.0     81.0      49.0       82.0        56.0   
3     43.0       31.0     67.0     43.0      64.0       12.0        38.0   
4     77.

## Prediction

Finalmente, ya tenemos el dataset construido con los datos que consideros mas relativos para la variable objetivo "Valor".

In [31]:
val = base.pop("Value")
base

Age  Overall  Potential      Wage  Special  International Reputation  \
0       31       94         94  565000.0     2202                       5.0   
1       33       94         94  405000.0     2228                       5.0   
2       26       92         93  290000.0     2143                       5.0   
3       27       91         93  260000.0     1471                       4.0   
4       27       91         92  355000.0     2281                       4.0   
...    ...      ...        ...       ...      ...                       ...   
18154   19       47         65    1000.0     1307                       1.0   
18155   19       47         63    1000.0     1098                       1.0   
18156   16       47         67    1000.0     1189                       1.0   
18157   17       47         66    1000.0     1228                       1.0   
18158   16       46         66    1000.0     1321                       1.0   

       Weak Foot  Skill Moves  Height  Weight      LS      ST      RS  \
0            4.0          4.0    5.70   159.0  90.000  90.000  90.000   
1            4.0          5.0    6.20   183.0  94.000  94.000  94.000   
2            5.0          5.0    5.90   150.0  87.000  87.000  87.000   
3            3.0          1.0    6.40   168.0  59.836  59.836  59.836   
4            5.0          4.0    5.11   154.0  85.000  85.000  85.000   
...          ...          ...     ...     ...     ...     ...     ...   
18154        2.0          2.0    5.90   134.0  44.000  44.000  44.000   
18155        2.0          2.0    6.30   170.0  47.000  47.000  47.000   
18156        3.0          2.0    5.80   148.0  47.000  47.000  47.000   
18157        3.0          2.0    5.10   154.0  49.000  49.000  49.000   
18158        3.0          2.0    5.10   176.0  45.000  45.000  45.000   

              LW         LF         CF         RF         RW       LAM  \
0      94.000000  95.000000  95.000000  95.000000  94.000000  95.00000   
1      92.000000  93.000000  93.000000  93.000000  92.000000  91.00000   
2      92.000000  92.000000  92.000000  92.000000  92.000000  92.00000   
3      61.058181  60.739921  60.739921  60.739921  61.058181  61.00366   
4      90.000000  90.000000  90.000000  90.000000  90.000000  91.00000   
...          ...        ...        ...        ...        ...       ...   
18154  46.000000  46.000000  46.000000  46.000000  46.000000  47.00000   
18155  41.000000  44.000000  44.000000  44.000000  41.000000  42.00000   
18156  47.000000  48.000000  48.000000  48.000000  47.000000  46.00000   
18157  49.000000  48.000000  48.000000  48.000000  49.000000  47.00000   
18158  47.000000  46.000000  46.000000  46.000000  47.000000  47.00000   

            CAM       RAM         LM        LCM         CM        RCM  \
0      95.00000  95.00000  93.000000  86.000000  86.000000  86.000000   
1      91.00000  91.00000  91.000000  84.000000  84.000000  84.000000   
2      92.00000  92.00000  91.000000  84.000000  84.000000  84.000000   
3      61.00366  61.00366  61.755241  60.241409  60.241409  60.241409   
4      91.00000  91.00000  91.000000  90.000000  90.000000  90.000000   
...         ...       ...        ...        ...        ...        ...   
18154  47.00000  47.00000  46.000000  47.000000  47.000000  47.000000   
18155  42.00000  42.00000  40.000000  37.000000  37.000000  37.000000   
18156  46.00000  46.00000  46.000000  40.000000  40.000000  40.000000   
18157  47.00000  47.00000  48.000000  41.000000  41.000000  41.000000   
18158  47.00000  47.00000  48.000000  47.000000  47.000000  47.000000   

              RM        LWB        LDM        CDM        RDM        RWB  \
0      93.000000  66.000000  63.000000  63.000000  63.000000  66.000000   
1      91.000000  68.000000  64.000000  64.000000  64.000000  68.000000   
2      91.000000  68.000000  63.000000  63.000000  63.000000  68.000000   
3      61.755241  59.533805  58.845677  58.845677  58.845677  59.533805   
4      91.000000  80.000000  80.000000  

Now we have only the club information, in hot-econding style. At the same time we have the data we want to predict in a numeric format, the value information. So now we can use this data to train a model. But first we need to split the data to being able to known its true performance.

In [32]:
X_train, X_test, y_train, y_test = train_test_split(base, val, test_size=0.33, random_state=42)

In [33]:
len(X_train)

12166

Now we train a LinearRegression model.

In [34]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

Finally we get a metric $R^2$ for the regression, we use the implementation from <a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html"> sickit-learn</a>.

In [35]:
preds = reg.predict(X_test)

In [36]:
preds[0]

707282.3607195723

In [37]:
y_test[0]

110500000.0

In [38]:
r2_score(y_test, preds)

0.9641759434065622

Obtenemos un 96% de fiabilidad, un valor muy cercano al 100%, lo cual nos demuestra que he contruido el dataset base con datos de entrenamiento relativos a la variable objetivo Valor